### ISSUE TITLE
Templates

### ISSUE THREAD

**Issue by AdamSpeight2008** [dotnet/vblang#68]
# Templates

### Synopsis

Simplify how code is constructed, via code.

```vbnet
Template DecalarationAndCreateInstance( ... )
  Dim ${0}$ As ${1}$ = New ${1}$(${2}$)
End Template
```
When invoked with the correct parameters, outputs as its result a `SyntaxNode` object model representing the code ascribed within the template.

```vbnet
 Dim thisNode = DeclarationAndCreateInstance( ... )
```

-----------------

### Internals

The compiler / ide produces on the coder's behalf a code-behind implementation. 
```vbnet
Private Class Templates

  Shared Function DeclarationAndCreateInstance( ) As SyntaxNode

  End Function

End Class
```
The internals of the method is could be generated via some mechanism like [RoslynQuoter](https://github.com/KirillOsenkov/RoslynQuoter). 

### Syntax Error and Highlighting
Within the template only basic syntax highlighting is provided, **No diagnostics** are supported. The template is provide "as seen / as is". *Note this doesn't rule out support for functionality at later date, just is won't be in this proposal.*

### Parsing
The parsing of the template *I think* would very similar to the parsing of Interpolation Strings.

**Comment by Bill-McC** [comment]
Would like to see more practical examples of how this templating would be used.

My preference would be to use snippets with better snippet tracking/integration into the IDE.  Developers could insert a snippet, and replace parameters as they do today and then right click on the snippet and select to "link snippet". The snippet linked with the code file and ideally also inserted into the project.

Advantages are : 
no new syntax; 
code re-use between projects using snippet library;
allows single use opt out of templating;
uses existing tools (in part);

**Comment by AdamSpeight2008** [comment]
@Bill-McC This proposal has nothing to do with snippets, if it were I'd have stated that.
It is about simplifying the generation of the code, as compared to manually produced written code using the Roslyn API (very verbose), to produce the code in the template.  If there currently tool that can generating it for you (ie RoslynQuoter), we should integrate it in to the language / IDE.

**Comment by Bill-McC** [comment]
@AdamSpeight2008   You asked for a template with parameters to simplify code construction: that has everything to do with snippets.  So I definitely wouldn't say it has "nothing" to do with snippets. Can you provide any examples where snippets would not achieve the stated goals you asked for ?

**Comment by xieguigang** [comment]
Yes, agree with you, but In my opinion, using your syntax probably too much complicated:

```vbnet
Template DecalarationAndCreateInstance( ... )
  Dim ${0}$ As ${1}$ = New ${1}$(${2}$)
End Template
```

And I think, declaring a template method can be introduced by add a new precompile keyword ``#Template .... #End Template``:

```vbnet
Public Module NumericSequence

#Template "Numeric sequence generator method"

    ' Template function declaring in template precompile keyword marked region
    Public Iterator Function Range(Of T)(from As T, [to] As T, Optional steps As T = 1) As IEnumerable(Of T)
        For i As T = from To [to] Step [step]
            Yield i
        Next
    End Function

    ' Due to the reason of the Declare keyword in VB.NET is using for the
    ' external C function call, but probably we Not Frequently using this keyword, 
    ' so I think we can using this Declare keyword for Template method instantiation 
    ' instantiation of the function from template
    Public Declare Function Range As Range(Of Integer)
    Public Declare Function Range As Range(Of Double)
    Public Declare Function Range As Range(Of Long)
    Public Declare Function Range As Range(Of Byte)
    Public Declare Function Range As Range(Of Short)
    Public Declare Function Range As Range(Of ULong)
    Public Declare Function Range As Range(Of UInteger)
    Public Declare Function Range As Range(Of BigInteger)

    ' or even alias name
    Public Declare Function VectorRange As Range(Of Vector)

#End Template

End Module

' Using the template function
Dim a% =1 
Dim b% = 100
Dim seq = Range(1, 100)

Dim x As Vector = {1, 2, 3, 4, 5}
Dim y As Vector = {100, 200, 300, 400, 500}
Dim seq As Matrix = VectorRange(x, y)
```

Using the template feature can save much of our coding work, in something like:

https://github.com/xieguigang/sciBASIC/blob/d9b13928df4df4dec3ef1f73d30cf9db2d93c4a0/Microsoft.VisualBasic.Architecture.Framework/Extensions/Collection/NumericSequence.vb

**Comment by AdamSpeight2008** [comment]
Both of you haven't got the point of this proposal, which has a very specific target, assisting the the development of code analyzers and code fixes. ([This stuff](https://github.com/dotnet/roslyn/wiki/Roslyn%20Overview)) A code analyser uses a Object Model represent of the code, not the textual *(ide)* representation. It is the production of that Object Model that represents your code isn't a simple as it code be.  

Template => Code To Generate code object at **runtime**.

**Comment by AdamSpeight2008** [comment]
Suppose you tasked to write an analyser and codefix that updates all 
```vbnet
ReadOnly Property ... () As ...
   Get
     Return ...
  End Get
End Property
```
to the following form.
```vb.net
ReadOnly Propery ... () As ... = ...
```

Using `Template` the core aspect, of find and replace in the code object model nodes, could be as simple as the following.
```vbnet
Template _ROProp0_
  Public ReadOnly ${Name}$() As ${[Type]}$
    Get
      Return ${Value}$
    End Get
 End Get
End Template
Template _ROProp1_
  Public ReadOnly ${Name}$() As ${[Type]) = ${Value}$
End Template

' Find all the "old style" properties.
Dim results As TemplateMatches = code.AllMatches( _ROProp )
' Replace them with the "new style" properties.
Dim newcode = results.ReplaceAllWith( _ROProp1_ )
```
Also with `template` you could embed the `Template` directly in the expressions, akin to lambda functions.

**Comment by Bill-McC** [comment]
@AdamSpeight2008  So are you asking for a template syntax in VB, that only produces outputs of SyntaxNode at runtime ?
In that case, that would be for only a few very specific tools and I don't think appropriate for a language inclusion. For your templates to be considered as a language addition they would need to have a much wider audience, IMO, as is what we typically see with "template" requests.

@xieguigang Your example brings to mind the INumeric requests that most of us have seen/wished for, at some stage over the last decade and a half.  (I still wonder if generics with a type constraint of{Integer, Double, Short, etc, etc} as would solve that. But more to the point, the problem with a template like you suggest is exactly that: the type constraint and when it gets compiled /verified.

**Comment by AdamSpeight2008** [comment]
@Bill-McC Yes  If you want compile time macros, you're welcome to create a proposal, but it likely to be shot down as it has been proposed multiple time and the language teams have declined.

**Comment by Bill-McC** [comment]
@AdamSpeight2008  No, not macros, templates ;)  And yes I know many of the things I read again and again have been asked before: the reasons they have not yet been implemented isn't always "no, never", and we've seen generics, snippets, and talk of mixins. The Grand Canyon didn't happen overnight, but it did happen.

In your scenario,  if you have your "templates", why can't you use the Roslyn Quoter or similar and then wrap it's output into factory methods that take the parameters you want replaced ? Seems simpler than getting the language actually changed. Should stage 1 be to support VB in the Roslyn Quoter ?

**Comment by Echo-8-ERA** [comment]
Isn't this from a conceptual POV a roundabout way of having a SyntaxNode version of XML Literals?

**Comment by AdamSpeight2008** [comment]
I think they are call Code Quotations in FSharp.

**Comment by Echo-8-ERA** [comment]
Admittedly, I have no idea how F# works, but between your examples earlier and the [Code Quotations documentation](https://docs.microsoft.com/en-us/dotnet/articles/fsharp/language-reference/code-quotations), this proposal really looks like it wants to be a "SystaxNode" Literal to me.

**Comment by AnthonyDGreen** [comment]
Code quotations are closer to VB/C# expression trees. When we first implemented interpolated string literals to be _target-typed_ to FormattableString we did consider a novel application could be something like this. I'd be curious to see if this could be implemented in that way.

That said for actual code generation we have a language-neutral code generation API in Roslyn that we prefer because it can target both VB and C# with the same code (which is a win for everybody).
